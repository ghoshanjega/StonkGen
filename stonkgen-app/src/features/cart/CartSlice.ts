import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import axios, { AxiosError } from "axios";
import { IBookCartItem, ICartItem } from "../../api-interface/Cart";
import { IApiError } from "../../api-interface/Error";
import { AppThunk, RootState } from "../../app/store";

export interface ICartItemExtended extends ICartItem {
  isLoading: boolean;
  displayStatus: string;
  displayPrice: number | null;
  amount: number | null;
  executionMode: string;
  selected: boolean;
  error: null | IApiError;
}

interface CartState {
  isLoading: boolean;
  error: AxiosError | null;
  cartItems: ICartItemExtended[] | null;
}

const initialState: CartState = {
  isLoading: true,
  error: null,
  cartItems: null,
};



export const cartSlice = createSlice({
  name: "cart",
  initialState,
  reducers: {
    fetchStarted: (state, action) => {
      state.isLoading = true;
    },
    fetchSuccessful: (state, action) => {
      state.isLoading = false;
      const cartItemsExtended = action.payload as ICartItemExtended[];
      state.cartItems = cartItemsExtended;
    },
    fetchUnSuccessful: (state, action) => {
      state.isLoading = false;
      state.error = action.payload;
    },
    setSelect: (state, action) => {
      const index = action.payload.index as number;
      if (state.cartItems) {
        state.cartItems[index].selected = action.payload.select;
      }
    },
    setExecutionMode: (state, action) => {
      const index = action.payload.index as number;
      if (state.cartItems) {
        state.cartItems[index].executionMode = action.payload.executionMode;
      }
    },
    setDisplayPrice: (state, action) => {
      const index = action.payload.index as number;
      if (state.cartItems) {
        state.cartItems[index].displayPrice = action.payload.orderPrice;
      }
    },
    setAmount: (state, action) => {
      const index = action.payload.index as number;
      if (state.cartItems) {
        state.cartItems[index].amount = action.payload.amount;
      }
    },
    setDisplayStatus: (state, action) => {
      if (state.cartItems) {
        const itemIndex = state.cartItems.findIndex((item) => item.id === action.payload.id);
        if (itemIndex > -1 && state.cartItems) {
          const item = state.cartItems[itemIndex]
          console.log("actions",action.payload)
          state.cartItems[itemIndex].displayStatus = computeDisplayStatus(item);
        }
      }
      // const index = action.payload.index as number;
      // if (state.cartItems) {
      //   const item = state.cartItems[index]
      //   state.cartItems[index].displayStatus = computeDisplayStatus(item);
      // }

    },
    removeItems: (state, action) => {
      const removeIds = action.payload.itemId as string[];
      if (state.cartItems) {
        state.cartItems = state.cartItems.filter((item) => !removeIds.includes(item.id));
      }
    },
    updateItem: (state, action) => {
      if (state.cartItems) {
        const itemIndex = state.cartItems.findIndex((item) => item.id === action.payload.id);
        if (itemIndex > -1 && state.cartItems) {
          state.cartItems[itemIndex] = {
            ...state.cartItems[itemIndex],
            ...action.payload,
          };
        }
      }
    },
  },
});

export const {
  fetchStarted,
  fetchSuccessful,
  fetchUnSuccessful,
  setExecutionMode,
  setSelect,
  setAmount,
  setDisplayPrice,
  setDisplayStatus,
  removeItems,
  updateItem,
} = cartSlice.actions;

const computeDisplayStatus = (item : ICartItemExtended) => {
  if (item.status === "prebooking") {
    if (
      item.executionSide &&
      item.executionMode &&
      item.stock.stockId &&
      item.stock.currency &&
      item.amount &&
      item.amount > 0
    ) {
      if (item.executionMode === "limit" && item.displayPrice && item.displayPrice > 0) {
        return "ready";
      } else if (item.executionMode === "market") {
        return "ready";
      }
      return "notReady";
    } else {
      return "notReady";
    }
  } else {
    return item.status;
  }
};

const computeExtended = (data: ICartItem[]): ICartItemExtended[] => {
  const itemsExtended: ICartItemExtended[] = data.map((item) => {
    return {
      ...item,
      isLoading: false,
      displayStatus: item.status === "prebooking" ? "notReady" : item.status,
      displayPrice: item.priceBooked ? item.priceBooked : null,
      executionMode: item.executionModeBooked ? item.executionModeBooked : "market",
      selected: false,
      amount: item.amountBooked ? item.amountBooked : null,
      error: null,
    };
  });
  return itemsExtended;
};

export const fetchCartItems = (params: any): AppThunk => (dispatch) => {
  dispatch(fetchStarted);
  return axios
    .get("http://localhost:3001/cart", {
      params,
    })
    .then(({ data }) => {
      dispatch(fetchSuccessful(computeExtended(data)));
    })
    .catch((err: AxiosError) => {
      dispatch(fetchUnSuccessful(err));
    });
};

export const removeCartItems = (params: any): AppThunk => (dispatch) => {
  const itemsToRemove = params.selected as ICartItemExtended[];
  return axios
    .patch("http://localhost:3001/cart/remove", {
      itemId: itemsToRemove.map((item) => item.id),
    })
    .then(({ data }) => {
      dispatch(removeItems({ itemId: itemsToRemove.map((item) => item.id) }));
    })
    .catch((err: AxiosError) => {
      dispatch(fetchUnSuccessful(err.response?.data));
    });
};

export const bookCartItem = (params: any): AppThunk => (dispatch) => {
  const item = params as ICartItemExtended;
  dispatch(updateItem({ id: item.id, selected: false,status: "inProgress" }));
  const body : IBookCartItem = {
    executionSide: item.executionSide, 
    executionMode: item.executionMode,
    displayPrice: item.displayPrice,
    amount: item.amount,
  }
  return axios
    .patch(`http://localhost:3001/cart/${item.id}/book`, body )
    .then(({ data }) => {
      dispatch(
        updateItem({
          id: data.id,
          status: data.status,
          displayStatus: data.status,
          displayPrice: data.priceBooked,
        })
      );
    })
    .catch((err: AxiosError) => {
      console.log("err", err.response);
      dispatch(
        updateItem({
          id: item.id,
          status: "rejected",
          displayStatus: "rejected",
          error: err.response?.data as IApiError,
        })
      );
    });
};

export const selectCartItems = (state: RootState) => state.cart.cartItems;

export default cartSlice.reducer;
