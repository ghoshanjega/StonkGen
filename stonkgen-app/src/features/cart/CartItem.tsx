import React from "react";

import { ICartItemExtended } from "./CartSlice";
import { DisplayStatus } from "./DisplayStatus";

interface Props {
  item: ICartItemExtended;
  handleItemChange: (payload: any, id: string, refreshStatus?: boolean) => void;
}

export const CartItem = ({ item, handleItemChange }: Props) => {
  const isOrderPriceDisabled = () => {
    if (
      item.executionMode === "market" ||
      ["inProgress", "booked", "rejected"].includes(item.displayStatus) ||
      item.status === "error"
    ) {
      return true;
    }
    return false;
  };
  const isAmountDisabled = () => {
    if (["inProgress", "booked", "rejected"].includes(item.displayStatus) || item.status === "error") {
      return true;
    }
    return false;
  };

  const isCheckBoxDisabled = () => {
    if (item.displayStatus === "inProgress") {
      return true;
    }
    return false;
  };

  if (item.error) {
    return (
      <tr>
        <td colSpan={12} style={{ borderRadius: "5px", backgroundColor: "#f5b0b3", color: "#753d3f" }}>
          <div className="d-flex justify-content-between align-items-center px-2">
            <span>
              <b>Error: </b>
              {item.error.message}
            </span>
            <button className="btn btn-light" onClick={() => handleItemChange({ error: null }, item.id,false)}>
              Close
            </button>
          </div>
        </td>
      </tr>
    );
  } else {
    return (
      <tr>
        {/* checkbox */}
        <td>
          <input
            type="checkbox"
            disabled={isCheckBoxDisabled()}
            checked={item.selected}
            onChange={(e) => handleItemChange({ selected: e.target.checked }, item.id)}
          />
        </td>
        {/* status */}
        <td>
          <DisplayStatus displayStatus={item.displayStatus} />{" "}
        </td>
        {/* execution side */}
        <td data-test="executionSide" style={{ textTransform: "capitalize" }}>{item.executionSide}</td>
        {/* stock code */}
        <td data-test="bloombergTickerLocal">{item.stock.bloombergTickerLocal}</td>
        {/* execution mode */}
        <td>
          <select
            className="form-control"
            data-test="executionMode"
            value={item.executionMode}
            disabled={isAmountDisabled()}
            onChange={(e) => handleItemChange({ executionMode: e.target.value }, item.id)}
          >
            <option value={"market"}>Market</option>
            <option value={"limit"}>Limit</option>
          </select>
        </td>
        {/* order price and currency */}
        <td>
          <div className="input-group">
            <input
              type="number"
              data-test="orderPrice"
              value={item.displayPrice || ""}
              className="form-control"
              disabled={isOrderPriceDisabled()}
              onChange={(e) => handleItemChange({ displayPrice: e.target.valueAsNumber }, item.id)}
            />
            <div className="input-group-append">
              <span className="input-group-text" data-test="currency">{item.stock.currency}</span>
            </div>
          </div>
        </td>
        {/* amount (shares) */}
        <td>
          <input
            type="number"
            value={item.amount || ""}
            data-test="amount"
            className="form-control"
            disabled={isAmountDisabled()}
            onChange={(e) => handleItemChange({ amount: e.target.valueAsNumber }, item.id)}
          />
        </td>
      </tr>
    );
  }
};
