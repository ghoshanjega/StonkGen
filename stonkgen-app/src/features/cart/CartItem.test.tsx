import React from "react";
import { CartItem } from "./CartItem";
import { ICartItemExtended } from "./CartSlice";
import { shallow } from 'enzyme';


const mockItem : ICartItemExtended = {
  stock :{
    stockId: "7f5f0946-1144-472a-b14e-61bb39a7d976",
    currency: "HKD",
    ric: "0434.HK",
    bloombergTicker: "434 HK",
    bloombergTickerLocal: "434 HK",
    name: "Boyaa Interactive International Ltd",
    country: "Hong Kong",
    price: 500.24
  },
  displayPrice : null,
  displayStatus: "notReady",
  executionMode: "market",
  executionSide: "buy",
  amount: null,
  error: null,
  status: "prebooking",
  id: "123123",
  selected: false,
  isLoading: false

}
const handleItemChange  = () => {}

describe("Cart Item Component", () => {
  
  it("should render the cart Item", () => {
    let component = shallow(<CartItem item={mockItem} handleItemChange={handleItemChange} />) 
    expect(component).toMatchSnapshot();
  });

  it("should match mock props", () => {
    let component = shallow(<CartItem item={mockItem} handleItemChange={handleItemChange} />) 
    expect(component.find('[data-test="bloombergTickerLocal"]').text()).toEqual(mockItem.stock.bloombergTickerLocal)
    expect(component.find('[data-test="executionSide"]').props().children).toEqual(mockItem.executionSide)
    expect(component.find('[data-test="orderPrice"]').props().value).toEqual("")
    expect(component.find('[data-test="currency"]').text()).toEqual(mockItem.stock.currency)
    expect(component.find('[data-test="amount"]').props().value).toEqual("")
    expect(component.find('[data-test="executionMode"]').props().value).toEqual(mockItem.executionMode)
  });

  it("should handle amount change", () => {
    const mockedFunction = jest.fn();
    let component = shallow(<CartItem item={mockItem} handleItemChange={mockedFunction} />) 
    const amountInput = component.find('[data-test="amount"]')
    expect(amountInput.props().value).toEqual("")
    amountInput.simulate('change',{ target: { valueAsNumber: 1000 } });
    expect(mockedFunction).toHaveBeenCalledWith({ amount: 1000},mockItem.id)
  });

  it("should handle executionMode change", () => {
    const mockedFunction = jest.fn();
    let component = shallow(<CartItem item={mockItem} handleItemChange={mockedFunction} />) 
    const amountInput = component.find('[data-test="executionMode"]')
    expect(amountInput.props().value).toEqual("market")
    amountInput.simulate('change',{ target: { value: "limit" } });
    expect(mockedFunction).toHaveBeenCalledWith({ executionMode: "limit"},mockItem.id)
  });

  it("should disable orderPrice on market execution mode", () => {
    const mockedFunction = jest.fn();
    let component = shallow(<CartItem item={mockItem} handleItemChange={mockedFunction} />) 
    const amountInput = component.find('[data-test="executionMode"]')
    expect(amountInput.props().value).toEqual("market")
    expect(component.find('[data-test="orderPrice"]').props().disabled).toEqual(true)
  });

  it("should enable orderPrice on limit execution mode", () => {
    const mockedFunction = jest.fn();
    let component = shallow(<CartItem item={{...mockItem, executionMode: "limit"}} handleItemChange={mockedFunction} />) 
    const amountInput = component.find('[data-test="executionMode"]')
    expect(amountInput.props().value).toEqual("limit")
    expect(component.find('[data-test="orderPrice"]').props().disabled).toEqual(false)
  });

  
});
