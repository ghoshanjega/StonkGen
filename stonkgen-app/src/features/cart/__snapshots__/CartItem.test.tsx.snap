// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart Item Component should render the cart Item 1`] = `
<tr>
  <td>
    <input
      checked={false}
      disabled={false}
      onChange={[Function]}
      type="checkbox"
    />
  </td>
  <td>
    <DisplayStatus
      displayStatus="notReady"
    />
     
  </td>
  <td
    data-test="executionSide"
    style={
      Object {
        "textTransform": "capitalize",
      }
    }
  >
    buy
  </td>
  <td
    data-test="bloombergTickerLocal"
  >
    434 HK
  </td>
  <td>
    <select
      className="form-control"
      data-test="executionMode"
      disabled={false}
      onChange={[Function]}
      value="market"
    >
      <option
        value="market"
      >
        Market
      </option>
      <option
        value="limit"
      >
        Limit
      </option>
    </select>
  </td>
  <td>
    <div
      className="input-group"
    >
      <input
        className="form-control"
        data-test="orderPrice"
        disabled={true}
        onChange={[Function]}
        type="number"
        value=""
      />
      <div
        className="input-group-append"
      >
        <span
          className="input-group-text"
          data-test="currency"
        >
          HKD
        </span>
      </div>
    </div>
  </td>
  <td>
    <input
      className="form-control"
      data-test="amount"
      disabled={false}
      onChange={[Function]}
      type="number"
      value=""
    />
  </td>
</tr>
`;
